### PR TITLE
add enhancement document for the feature `allow recurring backup for detached volumes`

### PR DESCRIPTION
**Pull request by PhanLe1010**
longhorn/longhorn#1509

### CODE REVIEW

File: enhancements/20201002-allow-recurring-backup-detached-volumes.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+https://github.com/longhorn/longhorn/issues/1509
+
+## Motivation
+
+### Goals
+
+1. Give the users an option to allow recurring backup to happen when the volume is detached.
+1. Don't overwrite the old backups with new identical backups. 
+   This happens when the volume is detached for a long time and has no new data.
+1. Avoid conflict with the data locality feature.
```
Comment (yasker) [inline enhancements/20201002-allow-recurring-backup-detached-volumes.md:22]: `if the automatically attach happens.`

File: enhancements/20201002-allow-recurring-backup-detached-volumes.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+### Non-goals
+
+1. The volume will not be available while the recurring backup process is happening.
+
+## Proposal
+
+1. Add new global boolean setting `Allow-Recurring-Backup-While-Volume-Detached`
```
Comment (yasker) [inline enhancements/20201002-allow-recurring-backup-detached-volumes.md:30]: Merge those two sentences.

File: enhancements/20201002-allow-recurring-backup-detached-volumes.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+#### Story 1
+
+Users use [OpenFaas framework](https://www.openfaas.com/). 
+OpenFaaS (Functions as a Service) is a framework for building serverless functions with Docker and Kubernetes.
+Users deploy serverless functions using Longhorn as persistent storage. 
+Users also set up recurring backup for those Longhorn volumes.
+OpenFass has APIGateway that watches and manages request to serverless functions.
+The nature of serverless functions is that the container that those functions run on are only created when APIGateway sees client requests to those functions.
+When there is no demand (e.g there is no function calls because there is no users' request), OpeenFaas will scale down the number of function replicas (similar to Pod concept in Kubernetes) to 0.
```
Comment (yasker) [inline enhancements/20201002-allow-recurring-backup-detached-volumes.md:43]: `OpenFaas`.

File: enhancements/20201002-allow-recurring-backup-detached-volumes.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+## Design
+
+### Implementation Overview
+
+1. Add new global boolean setting `Allow-Recurring-Backup-While-Volume-Detached`
+1. Modify the cronjob to do the following:
+   1. Check to see if the volume is attached.
+   1. If the volume is attached, we follow the same process as the current implementation.
+   1. If the volume is detached and `Allow-Recurring-Backup-While-Volume-Detached` is set to `true`, 
+      set `DataLocality` to `disabled` for the volume.
```
Comment (yasker) [inline enhancements/20201002-allow-recurring-backup-detached-volumes.md:77]: You cannot change the spec of the volume.

File: enhancements/20201002-allow-recurring-backup-detached-volumes.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+1. Modify the cronjob to do the following:
+   1. Check to see if the volume is attached.
+   1. If the volume is attached, we follow the same process as the current implementation.
+   1. If the volume is detached and `Allow-Recurring-Backup-While-Volume-Detached` is set to `true`, 
+      set `DataLocality` to `disabled` for the volume. 
+      We are about to do an auto attachment, and we don't want to create a new replica on the attached node.
+   1. Attach the volume to the node of the current Longhorn manager.
+      Also, disable the volume's frontend in the attaching request.
+      Disable the volume's frontend make sure that pod cannot use the volume during the recurring backup process.
+      This is necessary so that we can safely detach the volume when finishing the backup.
```
Comment (yasker) [inline enhancements/20201002-allow-recurring-backup-detached-volumes.md:82]: Disable Volume frontend should automatically disable the data locality.

File: enhancements/20201002-allow-recurring-backup-detached-volumes.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+1. Turn on data locality for the volume
+1. Let say the volume is attaching to node-1. 
+   Wait until there is a healthy replica on node-1.
+1. Disable scheduling for the node-1. This will force Kubernetes to avoid scheduling new pods on node-1.
+1. Write some data to the volume.
+1. Detach the volume
+1. Wait until the recurring backup starts.
+   Verify that Longhorn automatically attaches the volume, does backup, then detaches the volume.
+   However, there is no new replica is created.
+
```
Comment (yasker) [inline enhancements/20201002-allow-recurring-backup-detached-volumes.md:122]: Reset the option once the test is done.

File: enhancements/20201002-allow-recurring-backup-detached-volumes.md [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+1. On very subsequence 2 minutes, verify that Longhorn doesn't do backup.
+   The reason that Longhorn does not do backup is there is no new data.
+
+1. Turn on data locality for the volume
+1. Let say the volume is attaching to node-1. 
+   Wait until there is a healthy replica on node-1.
+1. Disable scheduling for the node-1. This will force Kubernetes to avoid scheduling new pods on node-1.
+1. Write some data to the volume.
+1. Detach the volume
+1. Wait until the recurring backup starts.
```
Comment (yasker) [inline enhancements/20201002-allow-recurring-backup-detached-volumes.md:119]: Make sure Longhorn automatically attach the volume to a node that doesn't have a replica and doesn't trigger the rebuild.

**Review by yasker — CHANGES_REQUESTED** [review]

File: enhancements/20201002-allow-recurring-backup-detached-volumes.md (hunk 1 above)
Comment (PhanLe1010) [reply enhancements/20201002-allow-recurring-backup-detached-volumes.md]: Thank you. Fixed

File: enhancements/20201002-allow-recurring-backup-detached-volumes.md (hunk 2 above)
Comment (PhanLe1010) [reply enhancements/20201002-allow-recurring-backup-detached-volumes.md]: Thank you. Fixed

File: enhancements/20201002-allow-recurring-backup-detached-volumes.md (hunk 3 above)
Comment (PhanLe1010) [reply enhancements/20201002-allow-recurring-backup-detached-volumes.md]: Thank you. Fixed

File: enhancements/20201002-allow-recurring-backup-detached-volumes.md (hunk 4 above)
Comment (PhanLe1010) [reply enhancements/20201002-allow-recurring-backup-detached-volumes.md]: Thank you. Fixed

File: enhancements/20201002-allow-recurring-backup-detached-volumes.md (hunk 5 above)
Comment (PhanLe1010) [reply enhancements/20201002-allow-recurring-backup-detached-volumes.md:81]: Thank you. Fixed

File: enhancements/20201002-allow-recurring-backup-detached-volumes.md (hunk 6 above)
Comment (PhanLe1010) [reply enhancements/20201002-allow-recurring-backup-detached-volumes.md:130]: Thank you. Fixed

File: enhancements/20201002-allow-recurring-backup-detached-volumes.md (hunk 7 above)
Comment (PhanLe1010) [reply enhancements/20201002-allow-recurring-backup-detached-volumes.md:125]: Thank you. Fixed

**Review by yasker — APPROVED** [review]

**Pull request merged**